docs/config.rst
Configuration
=============

Setup
-----

Create the tool's configuration directory::

    mkdir ~/.config/ocdskingfisher-process

Download the sample main configuration file::

    curl https://raw.githubusercontent.com/open-contracting/kingfisher-process/main/samples/config.ini -o ~/.config/ocdskingfisher-process/config.ini

Open the main configuration file at ``~/.config/ocdskingfisher-process/config.ini``, and follow the instructions below to update it.

PostgreSQL
----------

.. note::

   This step is required. All other steps are optional.

Configure the database connection settings:

.. code-block:: ini

   [DBHOST]
   HOSTNAME = localhost
   PORT = 5432
   USERNAME = ocdskingfisher
   PASSWORD = 
   DBNAME = ocdskingfisher

If you prefer not to store the password in ``config.ini``, you can use the `PostgreSQL Password File <https://www.postgresql.org/docs/11/libpq-pgpass.html>`__, ``~/.pgpass``, which overrides any password in ``config.ini``. Otherwise, if you used the same settings as in the examples during :doc:`requirements-install`, you only need to set ``PASSWORD`` above.

To override ``config.ini`` and/or ``.pgpass``, set the ``KINGFISHER_PROCESS_DB_URI`` environment variable. This is useful to temporarily use a different database than your default database. For example, in a bash-like shell::

    export KINGFISHER_PROCESS_DB_URI='postgresql://user:password@localhost:5432/dbname'

.. _config-logging:

Logging
-------

This tool uses the `Python logging module <https://docs.python.org/3/library/logging.html>`__. Loggers are in the ``ocdskingfisher`` namespace.

Logging from the :doc:`cli/index` can be configured with a ``~/.config/ocdskingfisher-process/logging.json`` file. To download the default configuration::

    curl https://raw.githubusercontent.com/open-contracting/kingfisher-process/main/samples/logging.json -o ~/.config/ocdskingfisher-process/logging.json

To download a different configuration that includes debug messages::

    curl https://raw.githubusercontent.com/open-contracting/kingfisher-process/main/samples/logging-debug.json -o ~/.config/ocdskingfisher-process/logging.json

Web API
-------

To allow access to the :ref:`web API <web-api>`, set API keys, separated by commas. For example, to set ``1234`` and ``5678`` as keys (in practice, you should use `long, random keys <https://www.avast.com/en-us/random-password-generator>`__):

.. code-block:: ini

    [WEB]
    API_KEYS = 1234,5678

To override ``config.ini``, set the ``KINGFISHER_PROCESS_WEB_API_KEYS`` environment variable.

Collection flags
----------------

When a :ref:`new collection <collections>` is created, :ref:`flags <schema-check-flags>` are set to indicate what operations to perform on them.

All flags are off by default. To turn any on:

.. code-block:: ini

    [COLLECTION_DEFAULT]
    CHECK_DATA = true
    CHECK_OLDER_DATA_WITH_SCHEMA_1_1 = false

Default pre-processing pipeline
-------------------------------

To enable the :doc:`standard-pipeline`:

.. code-block:: ini

    [STANDARD_PIPELINE]
    RUN = true

Redis
-----

Configure the Redis connection settings:

.. code-block:: ini

    [REDIS]
    HOST = localhost
    PORT = 6379
    DATABASE = 0

Sentry
------

To track crashes, `sign up <https://sentry.io/signup/>`__ for `Sentry <https://sentry.io/>`__, and set the DSN:

.. code-block:: ini

    [SENTRY]
    DSN = https://<key>@sentry.io/<project>

.. note::

    Sentry has its own `environment variables <https://docs.sentry.io/error-reporting/configuration/?platform=python>`__. Further reading: `Sentry for Python <https://sentry.io/for/python/>`__.


ocdskingfisherprocess/config.py
import configparser
import os
import sys

import pgpasslib


"""This holds configuration information for Kingfisher.
Whatever tool is calling it - CLI or other code - should create one of these, set it up as required and pass it around.
"""


class Config:

    def __init__(self):
        self.web_api_keys = []
        self.database_uri = ''
        self._database_host = ''
        self._database_port = 5432
        self._database_user = ''
        self._database_name = ''
        self._database_password = ''
        self.run_standard_pipeline = False
        self.redis_host = ''
        self.redis_port = 6379
        self.redis_database = 0
        self.sentry_dsn = ''

    def load_user_config(self):
        # First, try and load any config in the ini files
        self._load_user_config_ini()
        # Second, loook for password in .pggass
        self._load_user_config_pgpass()
        # Third, try and load any config in the env (so env overwrites ini)
        self._load_user_config_env()

    def _load_user_config_pgpass(self):
        if not self._database_name or not self._database_user:
            return

        try:
            password = pgpasslib.getpass(
                self._database_host,
                self._database_port,
                self._database_name,
                self._database_user
            )
            if password:
                self._database_password = password
                self.database_uri = 'postgresql://{}:{}@{}:{}/{}'.format(
                    self._database_user,
                    self._database_password,
                    self._database_host,
                    self._database_port,
                    self._database_name
                )

        except pgpasslib.FileNotFound:
            # Fail silently when no files found.
            return
        except pgpasslib.InvalidPermissions:
            print(
                "Your pgpass file has the wrong permissions, for your safety this file will be ignored. " +
                "Please fix the permissions and try again.")
            return
        except pgpasslib.PgPassException:
            print("Unexpected error:", sys.exc_info()[0])
            return

    def _load_user_config_env(self):
        if os.environ.get('KINGFISHER_PROCESS_WEB_API_KEYS'):
            self.web_api_keys = [key.strip() for key in os.environ.get('KINGFISHER_PROCESS_WEB_API_KEYS').split(',')]

        if os.environ.get('KINGFISHER_PROCESS_DB_URI'):
            self.database_uri = os.environ.get('KINGFISHER_PROCESS_DB_URI')

    def _load_user_config_ini(self):
        config = configparser.ConfigParser()

        if os.path.isfile(os.path.expanduser('~/.config/ocdskingfisher-process/config.ini')):
            config.read(os.path.expanduser('~/.config/ocdskingfisher-process/config.ini'))
        else:
            return

        self.web_api_keys = [key.strip() for key in config.get('WEB', 'API_KEYS', fallback='').split(',')]

        self._database_host = config.get('DBHOST', 'HOSTNAME')
        self._database_port = config.get('DBHOST', 'PORT')
        self._database_user = config.get('DBHOST', 'USERNAME')
        self._database_name = config.get('DBHOST', 'DBNAME')
        self._database_password = config.get('DBHOST', 'PASSWORD', fallback='')

        self.database_uri = 'postgresql://{}:{}@{}:{}/{}'.format(
            self._database_user,
            self._database_password,
            self._database_host,
            self._database_port,
            self._database_name
        )

        self.run_standard_pipeline = \
            config.getboolean('STANDARD_PIPELINE', 'RUN', fallback=False)

        self.redis_host = config.get('REDIS', 'HOST', fallback='')
        self.redis_port = config.get('REDIS', 'PORT', fallback=6379)
        self.redis_database = config.get('REDIS', 'DATABASE', fallback=0)

        self.sentry_dsn = config.get('SENTRY', 'DSN', fallback='')

    def is_redis_available(self):
        return self.redis_host and self.redis_port
